### PR TITLE
feat(toolchain, pip.parse): introduce a new '_host' toolchain repo

### PR DESCRIPTION
**Pull request by aignas**
This is for passing it in repository_rules and relies on the canonical
label representation introduced in bazel 6.0 and symlink support (needs
to be present on Windows) to work.

This allows the users to not need to `load` the interpreter label from
a `.bzl` file but instead specify the label in the form of
`@<mytoolchainrepo>_host//:python`.

In order to make it work robustly on Windows, we do
`repository_ctx.path(python_interpreter_label).realpath` to get the actual
path of the Python interpreter on both, bzlmod and non-bzlmod setups on
Windows within the `whl_library` repository rule. 

Work towards #1643.

### CODE REVIEW

**Comment by rickeylev** [comment]
Oh this is gonna be tough to remember. @chrislovecnm do you remember what the issue was with windows and trying to symlink to the python interpreter from another repo?

**Comment by rickeylev** [comment]
Ah, this might be it: https://github.com/bazelbuild/rules_python/pull/1265

Something about: the symlink can't be executed directly / powershell issue / dll's can't be found from the symlink's location ?

**Comment by aignas** [comment]
I am wondering then why the CI is still passing. Maybe the
`repository_ctx` usage is good enough with `symlinks`, because
the `symlink` gets always resolve the real path before executing?

Just to be explicit, I am not using the symlink in toolchain
resolution. For that we already have the `alias` repos and this
is only for the `repository_ctx` usecase and the Windows CI seems
to be still happy. Do we need to make this `host_python` repo
internal to `rules_python` in some way?

**Comment by aignas** [comment]
Yeah, the previous `path.realpath` usage was only on Windows bzlmod, when I enabled it for everything, it started working again (the multi-python-version tests on non-bzlmod started failing because the `path.realpath` workaround was missing there.

File: CHANGELOG.md [hunk 1]
```diff
@@ -58,13 +58,23 @@ A brief description of the categories of changes:
 * (coverage): coverage reports are now created when the version-aware
   rules are used.
   ([#1600](https://github.com/bazelbuild/rules_python/issues/1600))
+* (bzlmod pip.parse) Use the same interpreter label that may make the lock file
+  almost the same for all platforms. It will, however be different in cases where
+  the lock file is introducing platform-specific deps (e.g. `colorama` may be
```
Comment (rickeylev) [inline CHANGELOG.md:63]: >  It will, however be different in cases where the lock file is introducing platform-specific deps

I think "it" here is referring to "when there are platform-specific deps"? As written, it sounds like it is referring to "interpreter label" (that the interpreter label will be different if you have platform-specific deps)

Maybe rephrase to something like:
> Use a platform-independent reference to the interpreter pip uses. This reduces (but doesn't eliminate) the amount of platform-specific content in MODULE lock files; Follow 1643 for removing platform-specific content in MODULE lock files

File: examples/pip_parse_vendored/BUILD.bazel [hunk 2]
```diff
@@ -35,14 +17,14 @@ write_file(
         "#!/usr/bin/env bash",
         # Bazel gives us a way to access the source folder!
         "cd $BUILD_WORKSPACE_DIRECTORY",
-        "cp -fv bazel-bin/requirements.clean.bzl requirements.bzl",
+        "cp -fv bazel-pip_parse_vendored/external/pip/requirements.bzl requirements.bzl",
```
[the remark this answers is not in the document]
Comment (aignas) [reply examples/pip_parse_vendored/BUILD.bazel:20]: Is the flag that you have in mind mentioned here? https://github.com/bazelbuild/bazel/issues/12821

File: examples/multi_python_versions/WORKSPACE [hunk 3]
```diff
@@ -28,19 +28,15 @@ python_register_multi_toolchains(
...
 
 multi_pip_parse(
     name = "pypi",
     default_version = default_python_version,
     python_interpreter_target = {
-        "3.10": interpreter_3_10,
-        "3.11": interpreter_3_11,
-        "3.8": interpreter_3_8,
-        "3.9": interpreter_3_9,
+        "3.10": "@python_3_10_host//:python",
```
[the remark this answers is not in the document]
Comment (aignas) [reply examples/multi_python_versions/WORKSPACE:36]: Yeah, once I started moving the legacy `WORKSPACE` examples since I could do this everything started falling into place pretty neatly.

File: examples/pip_parse_vendored/BUILD.bazel (hunk 2 above)
[the remark this answers is not in the document]
Comment (rickeylev) [reply examples/pip_parse_vendored/BUILD.bazel:20]: Yeah, lets just leave it. There saw some internal discussion about this, but it didn't sound conclusive. If it does change, then we'll find out via the downstream pipeline and PRs anyways.

**Review by rickeylev — APPROVED** [review]
